In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('us_data_with_dates.csv')
df['date'] = pd.to_datetime(df['date'])

# Get Rep Risk Rating with ISIN

In [14]:
import wrds

In [370]:
conn = wrds.Connection(wrds_username='jongmok1031')

Loading library list...
Done


In [35]:
query = conn.raw_sql("""select ri.*, ci.primary_ISIN, ci.name
                        from reprisk.pm_rri_data ri
                        left join reprisk.pm_company_identifiers ci
                        on ri.reprisk_id = ci.reprisk_id
                        where ci.headquarter_country = 'United States of America' and ri.date > '2020-01-01' and ci.primary_isin is not null
                        """, date_cols=['date'])
US_companies = pd.DataFrame(data=query)

## mapping of ISIN to ticker

In [85]:
isins = list(US_companies['primary_isin'].unique())

In [309]:
def get_us_ticker_form_isin(isin):

    # https://www.openfigi.com/api
    import requests

    url = 'https://api.openfigi.com/v1/mapping'
    headers = {'Content-Type':'text/json', 'X-OPENFIGI-APIKEY':'46a82b66-6569-4077-b14a-2aa771bb933f' }
    payload = '[{"idType":"ID_ISIN","idValue":"' + isin + '"}]'

    r = requests.post(url, headers=headers, data=payload)

    return r.json()[0].get('data')[0].get('ticker')

In [104]:
from time import sleep

# getting ISIN and returns by tickers

In [145]:
sp500_info = pd.read_html('https://en.wikipedia.org/wiki/List_of_S%26P_500_companies')[0]
sp500_symbols = list(df['Symbol'].values)

In [324]:
def get_isin_by_ticker(ticker):    
    response = requests.post(url='https://api.openfigi.com/v3/search/',
                            headers={'Content-Type':'text/json', 'X-OPENFIGI-APIKEY':'46a82b66-6569-4077-b14a-2aa771bb933f' },
                            json={
                                'query': ticker,
                                'exchCode': 'GM'
                            })
    try:
        request = requests.get(f"https://data.lemon.markets/v1/instruments/?search="+ response.json().get('data')[0].get('ticker')+"&type=stock",
                           headers={"Authorization": "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJsZW1vbi5tYXJrZXRzIiwiaXNzIjoibGVtb24ubWFya2V0cyIsInN1YiI6InVzcl9xeUdHQkREZmZIQ1hjUzhIYmJZSlM3MG1qS3BQNW15blZsIiwiZXhwIjoxNjgwNTY2NTQ2LCJpYXQiOjE2NDkwMzA1NDYsImp0aSI6ImFwa19xeUdHQkREaGg1S3dMV0s2YllnTTRSTTJWNHhuNVJyaktDIiwibW9kZSI6InBhcGVyIn0.5KgttSY7eWUDyYvQ_B9muWnMFcZJaOZlfnYyVNixVzM"})
        try:
            isin = request.json().get('results')[0].get('isin')
            return isin, response
        except:
            return 'fail', response
    except:
        return 'fail', response
        
     

In [325]:
isins = []
temp_r = ''
for i in range(len(sp500_symbols)):
    isin, response = get_isin_by_ticker(sp500_symbols[i])
    sleep(3.1)
    isins.append(isin)
    print(isin, sp500_symbols[i], response)
    if i%20 == 0 and i != 0:
        print('20 requests' + str(len(sp500_symbols)-i) + ' left' )

US88579Y1010 MMM <Response [200]>
GB0022569080 AOS <Response [200]>
CA00386B1094 ABT <Response [200]>
US00287Y1091 ABBV <Response [200]>
US0036541003 ABMD <Response [200]>
IE00B4BNMY34 ACN <Response [200]>
US00507V1098 ATVI <Response [200]>
US0394831020 ADM <Response [200]>
US00724F1012 ADBE <Response [200]>
US0530151036 ADP <Response [200]>
DE000A3H2101 AAP <Response [200]>
US00130H1059 AES <Response [200]>
US0010551028 AFL <Response [200]>
fail A <Response [200]>
US0268747849 AIG <Response [200]>
US0091581068 APD <Response [200]>
US00971T1016 AKAM <Response [200]>
US0116591092 ALK <Response [200]>
fail ALB <Response [200]>
US0152711091 ARE <Response [200]>
US0162551016 ALGN <Response [200]>
20 requests485 left
fail ALLE <Response [200]>
fail LNT <Response [200]>
fail ALL <Response [200]>
US02079K3059 GOOGL <Response [200]>
US02079K1079 GOOG <Response [200]>
US02209S1033 MO <Response [200]>
US0231351067 AMZN <Response [200]>
JE00BJ1F3079 AMCR <Response [200]>
US0079031078 AMD <Respons

US44107P1049 HST <Response [200]>
US4432011082 HWM <Response [200]>
20 requests265 left
US40434L1052 HPQ <Response [200]>
US4448591028 HUM <Response [200]>
US4464131063 HII <Response [200]>
US4461501045 HBAN <Response [200]>
US45167R1041 IEX <Response [200]>
US45168D1046 IDXX <Response [200]>
US4523081093 ITW <Response [200]>
US4523271090 ILMN <Response [200]>
US45337C1027 INCY <Response [200]>
IE00BK9ZQ967 IR <Response [200]>
US4581401001 INTC <Response [200]>
US45866F1049 ICE <Response [200]>
CA45921J1093 IBM <Response [200]>
fail IP <Response [200]>
US44980X1090 IPG <Response [200]>
US4595061015 IFF <Response [200]>
US4612021034 INTU <Response [200]>
US46120E6023 ISRG <Response [200]>
fail IVZ <Response [200]>
fail IPGP <Response [200]>
20 requests245 left
US46266C1053 IQV <Response [200]>
US46284V1017 IRM <Response [200]>
US4456581077 JBHT <Response [200]>
US4262811015 JKHY <Response [200]>
US4660321096 J <Response [200]>
US4781601046 JNJ <Response [200]>
IE00BY7QL619 JCI <Response

US9182041080 VFC <Response [200]>
US92556D1063 VTRS <Response [200]>
US92826C8394 V <Response [200]>
fail VNO <Response [200]>
20 requests25 left
US9291601097 VMC <Response [200]>
US9297401088 WAB <Response [200]>
US9311421039 WMT <Response [200]>
US9314271084 WBA <Response [200]>
fail WM <Response [200]>
US9418481035 WAT <Response [200]>
US92939U1060 WEC <Response [200]>
CA9528451052 WFC <Response [200]>
CA94947L1022 WELL <Response [200]>
US9553061055 WST <Response [200]>
US9581021055 WDC <Response [200]>
fail WRK <Response [200]>
US9621661043 WY <Response [200]>
US9633201069 WHR <Response [200]>
US9694571004 WMB <Response [200]>
fail WTW <Response [200]>
US9831341071 WYNN <Response [200]>
US98389B1008 XEL <Response [200]>
US98419M1009 XYL <Response [200]>
US9884981013 YUM <Response [200]>
20 requests5 left
US9892071054 ZBRA <Response [200]>
IL0065100930 ZBH <Response [200]>
US9896961094 ZION <Response [200]>
US98978V1035 ZTS <Response [200]>


In [326]:
isins_ticker_sp500 = pd.DataFrame([isins, sp500_symbols]).T
isins_ticker_sp500.columns = ['primary_isin', 'ticker_US']

In [329]:
#isins_ticker_sp500.to_csv('isins_ticker_sp500_v2.csv')

In [330]:
isins_ticker_sp500['primary_isin'].value_counts()

fail            48
US92343V1044     2
US3755581036     2
US9139031002     1
IE00BLS09M33     1
                ..
US57667L1070     1
FR0000120222     1
US30212P3038     1
AN8068571086     1
US55261F1049     1
Name: primary_isin, Length: 456, dtype: int64

In [153]:
def get_isin_by_ticker_2(ticker):

    # https://www.openfigi.com/api
    import requests

    url = 'https://api.openfigi.com/v1/mapping'
    headers = {'Content-Type':'text/json', 'X-OPENFIGI-APIKEY':'46a82b66-6569-4077-b14a-2aa771bb933f' }
    payload = '[{"idType":"BASE_TICKER", "idValue":"' + ticker + '"}]'
    
    r = requests.post(url, headers=headers, data=payload)

    return r.json()

In [301]:
isins_ticker_sp500

primary_isin ticker_US
0    US88579Y1010       MMM
1    GB0022569080       AOS
2    CA00386B1094       ABT
3    US00287Y1091      ABBV
4    US0036541003      ABMD
..            ...       ...
500  US9884981013       YUM
501  US9892071054      ZBRA
502  IL0065100930       ZBH
503  US9896961094      ZION
504  US98978V1035       ZTS

[505 rows x 2 columns]

# Joining ticker data on rep risk data

In [331]:
merge1 = US_companies.merge(isins_ticker_sp500, how = 'inner', on = 'primary_isin')

In [332]:
merge1['primary_isin'].nunique()

328

In [333]:
merge2 = US_companies.merge(isins_ticker_sp500, how = 'right', on = 'primary_isin')

In [354]:
isins_ticker_sp500['primary_isin'].nunique()

456

In [356]:
merge2[merge2['name'].isna()]['primary_isin'].unique()

array(['GB0022569080', 'CA00386B1094', 'IE00B4BNMY34', 'DE000A3H2101',
       'fail', 'US0152711091', 'US02079K1079', 'JE00BJ1F3079',
       'AU000000AMP6', 'FR0004040608', 'NL0000888691', 'US03209T1097',
       'US03662Q1058', 'IE00BLP1HW54', 'US03743Q1085', 'JE00B783TY65',
       'US0404131064', 'CA0019401052', 'US0846707026', 'US03615A1088',
       'US09073M1045', 'AU0000164139', 'ES0113860A34', 'CH0006539198',
       'US1273871087', 'BE0003593044', 'GB0031215220', 'US1488061029',
       'FR0000120222', 'CA1421731037', 'CNE100001T80', 'CA22587M1068',
       'DE000A288904', 'US21037T1097', 'US2172041061', 'GB00B64NSP76',
       'ID1000115306', 'DE0005545503', 'US30711Y3009', 'US25470F3029',
       'CA25552L2012', 'AU000000DOW2', 'DK0061277894', 'IE00B8KQN827',
       'CA29787F1045', 'FR0013344173', 'US30212W1009', 'FR0000121667',
       'CH0126881561', 'LU0472835155', 'JP3802300008', 'CA32016U2074',
       'US33813J1060', 'AU000000FLT9', 'US34964C1062', 'CA7815861025',
       'AU0000

In [336]:
get_us_ticker_form_isin('BE0003593044')

'COFB'

In [335]:
isins_ticker_sp500[isins_ticker_sp500['primary_isin'] == 'BE0003593044']

primary_isin ticker_US
85  BE0003593044       COF

In [366]:
for i in US_companies['name']:
    if i.lower().find('prudential') > 0:
        print(i)

# getting reprisk inndex for s&p 500 companies (453 company ISIN's found so far, among them reprisk has info on only 328)

In [401]:
repriskdata = pd.DataFrame(columns = repriskdata.columns)
for i in range(isins_ticker_sp500.shape[0]):
    isin = "'"+isins_ticker_sp500.loc[i]['primary_isin']+"'"
    ticker_US = isins_ticker_sp500.loc[i]['ticker_US']
    if isin != 'fail':
        query = conn.raw_sql("""select ri.*, ci.primary_ISIN, ci.name
                                from reprisk.pm_rri_data ri
                                left join reprisk.pm_company_identifiers ci
                                on ri.reprisk_id = ci.reprisk_id
                                where ci.headquarter_country = 'United States of America' and ci.primary_isin is not null and ci.primary_ISIN = """ + isin, date_cols=['date'])
        repriskdata_t = pd.DataFrame(data=query)
        repriskdata_t['ticker'] = ticker_US
        repriskdata = pd.concat([repriskdata,repriskdata_t])

/Users/saveliykholin/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:13: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  del sys.path[0]


In [404]:
#repriskdata.to_csv('repriskdata_with_tickers_328_sp500.csv')

## getting monthly returns of sp500 compannies

In [408]:
import pandas_datareader as web

In [461]:
m_returns = pd.DataFrame(columns = m_returns.columns)
for i in range(isins_ticker_sp500.shape[0]):
    ticker_US = isins_ticker_sp500.loc[i]['ticker_US']
    try:
        returns = web.get_data_yahoo(ticker_US,
                            start = "2006-12-29",
                            end = "2020-12-31") # dates of reprisk data
        monthly_returns_t = returns['Adj Close'].resample('M').ffill().pct_change()
        monthly_returns_t = pd.DataFrame(monthly_returns_t[1:]).reset_index()
        monthly_returns_t['ticker'] = ticker_US
        monthly_returns_t.columns = ['date', 'monthly_return', 'ticker']
    except:
        monthly_returns_t['ticker'] = ticker_US
        monthly_returns_t['monthly_return'] = 'fail'
    
    m_returns = pd.concat([m_returns,monthly_returns_t])
    if i%20 == 0 and i != 0:
        print('20 requests, ' + str(isins_ticker_sp500.shape[0]-i) + ' left' )

20 requests, 485 left
20 requests, 465 left
20 requests, 445 left
20 requests, 425 left
20 requests, 405 left
20 requests, 385 left
20 requests, 365 left
20 requests, 345 left
20 requests, 325 left
20 requests, 305 left
20 requests, 285 left
20 requests, 265 left
20 requests, 245 left
20 requests, 225 left
20 requests, 205 left
20 requests, 185 left
20 requests, 165 left
20 requests, 145 left
20 requests, 125 left
20 requests, 105 left
20 requests, 85 left
20 requests, 65 left
20 requests, 45 left
20 requests, 25 left
20 requests, 5 left


In [463]:
#m_returns.to_csv('m_returns_sp500_companies_2007_2020.csv')

# creating master table with reprisk index and  returns

In [470]:
master_table = repriskdata.merge(m_returns, how = 'left', on = ['ticker', 'date'])

In [472]:
#master_table.to_csv('master_table_330comp_sp500.csv') 